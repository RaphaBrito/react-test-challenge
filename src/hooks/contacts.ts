import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import type { Contact, ContactFormData } from "../types/contact";

export function useContacts() {
  const {
    data: contacts,
    isPending,
    isError,
  } = useQuery<Contact[]>({
    queryKey: ["contacts"],
    queryFn: async () => {
      return (await fetch("http://localhost:5432/contacts")).json();
    },
  });

  return { contacts, isPending, isError };
}

export function useContactsCreateMutation() {
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (formData: ContactFormData) => {
      return fetch(`http://localhost:5432/contacts`, {
        method: "POST",
        body: JSON.stringify(formData),
        headers: {
          "Content-Type": "application/json",
        },
      });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["contacts"] });
    },
  });

  return mutation;
}

export function useContactsEditMutation() {
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (contact: Contact) => {
      return fetch(`http://localhost:5432/contacts/${contact.id}`, {
        method: "PUT",
        body: JSON.stringify(contact),
        headers: {
          "Content-Type": "application/json",
        },
      });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["contacts"] });
    },
  });

  return mutation;
}

export function useContactsDeleteMutation() {
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (id: number) => {
      return fetch(`http://localhost:5432/contacts/${id}`, {
        method: "DELETE",
      });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["contacts"] });
    },
  });

  return mutation;
}
