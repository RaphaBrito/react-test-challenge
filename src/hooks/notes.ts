import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import type { Note, NoteFormData } from "../types/note";

export function useNotes() {
  const {
    data: notes,
    isPending,
    isError,
  } = useQuery<Note[]>({
    queryKey: ["notes"],
    queryFn: async () => {
      return (await fetch("http://localhost:5432/notes")).json();
    },
  });

  return { notes, isPending, isError };
}

export function useNotesCreateMutation() {
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (formData: NoteFormData) => {
      return fetch(`http://localhost:5432/notes`, {
        method: "POST",
        body: JSON.stringify(formData),
        headers: {
          "Content-Type": "application/json",
        },
      });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["notes"] });
    },
  });

  return mutation;
}

export function useNotesEditMutation() {
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (note: Note) => {
      return fetch(`http://localhost:5432/notes/${note.id}`, {
        method: "PUT",
        body: JSON.stringify(note),
        headers: {
          "Content-Type": "application/json",
        },
      });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["notes"] });
    },
  });

  return mutation;
}

export function useNotesDeleteMutation() {
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (id: number) => {
      return fetch(`http://localhost:5432/notes/${id}`, {
        method: "DELETE",
      });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["notes"] });
    },
  });

  return mutation;
}
