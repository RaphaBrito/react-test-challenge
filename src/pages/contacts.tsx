import { AppError } from "../components/app-error";
import { ContactCard } from "../components/contact-card";
import { CreateContactCard } from "../components/create-contact-card";
import { Loading } from "../components/loading";
import {
  useContacts,
  useContactsCreateMutation,
  useContactsDeleteMutation,
  useContactsEditMutation,
} from "../hooks/contacts";
import styles from "../styles/list.module.css";

import type { Contact, ContactFormData } from "../types/contact";

export function Contacts() {
  const { contacts = [], isPending, isError } = useContacts();
  const createMutation = useContactsCreateMutation();
  const editMutation = useContactsEditMutation();
  const deleteMutation = useContactsDeleteMutation();

  const handleCreate = (formData: ContactFormData) => {
    createMutation.mutate(formData);
  };

  const handleEdit = (contact: Contact) => {
    editMutation.mutate(contact);
  };

  const handleDelete = (id: number) => {
    deleteMutation.mutate(id);
  };

  if (isPending) {
    return <Loading />;
  }

  if (isError) {
    return <AppError />;
  }

  return (
    <div>
      <h1>Contatos</h1>
      <div className={styles.listGrid}>
        {contacts.map((contact: Contact) => (
          <ContactCard
            key={contact.id}
            contact={contact}
            onDelete={handleDelete}
            onEdit={handleEdit}
          />
        ))}

        <CreateContactCard onCreate={handleCreate} />
      </div>
    </div>
  );
}
