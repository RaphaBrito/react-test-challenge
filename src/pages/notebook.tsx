import { AppError } from "@/components/app-error";
import { CreateNoteCard } from "@/components/create-note-card";
import { Loading } from "@/components/loading";
import { NoteCard } from "@/components/note-card";
import {
  useNotes,
  useNotesCreateMutation,
  useNotesDeleteMutation,
  useNotesEditMutation,
} from "@/hooks/notes";
import styles from "@/styles/list.module.css";

import type { Note, NoteFormData } from "@/types/note";

export function Notebook() {
  const { notes = [], isPending, isError } = useNotes();
  const createMutation = useNotesCreateMutation();
  const editMutation = useNotesEditMutation();
  const deleteMutation = useNotesDeleteMutation();

  const handleCreate = (formData: NoteFormData) => {
    createMutation.mutate(formData);
  };

  const handleEdit = (note: Note) => {
    editMutation.mutate(note);
  };

  const handleDelete = (id: number) => {
    deleteMutation.mutate(id);
  };

  if (isPending) {
    return <Loading />;
  }

  if (isError) {
    return <AppError />;
  }

  return (
    <div>
      <h1>Bloco de notas</h1>
      <div className={styles.listGrid}>
        {notes.map((note: Note) => (
          <NoteCard
            key={note.id}
            note={note}
            onDelete={handleDelete}
            onEdit={handleEdit}
          />
        ))}

        <CreateNoteCard onCreate={handleCreate} />
      </div>
    </div>
  );
}
