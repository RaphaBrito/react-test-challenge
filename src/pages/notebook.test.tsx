import { render, waitFor, screen } from "@testing-library/react";

import * as notes from "@/hooks/notes";
import { Notebook } from "@/pages/notebook";
import { NoteFormData, type Note } from "@/types/note";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import jsonServer from "json-server";
import { queryClient } from "@/services/queryClient";
import userEvent from "@testing-library/user-event";

const json = {
  contacts: [
    {
      id: 1,
      name: "Raphael",
      email: "brito@gmail.com",
      phone: "+558190836728",
    },
    {
      id: 2,
      name: "Cesar",
      email: "cesar@gmail.com",
      phone: "+558190836728",
    },
    {
      id: 3,
      name: "cesare",
      email: "Cesare@gmail.com",
      phone: "+558190836728",
    },
  ],
  notes: [
    {
      id: 1,
      title: "Lembrete",
      description: "Colocar presença na aula.",
    },
    {
      id: 2,
      title: "React",
      description: "Uma das bibliotecas mais utilizadas da atualidade.",
    },
    {
      id: 3,
      title: "CESAR",
      description: "Centro de Estudos e Sistemas Avançados do Recife.",
    },
  ],
};
// vi.mock("notes");

const server = jsonServer.create();
const router = jsonServer.router(json);
const middlewares = jsonServer.defaults();

server.use(middlewares);
server.use(router);
server.listen(5432, () => {
  console.log("JSON Server is running");
});

/*const mockedQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});*/

const ProviderWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

describe("Notebook component", () => {
  const mockNotes: Note[] = [
    { id: 1, title: "Lembrete", description: "Colocar presença na aula." },
    {
      id: 2,
      title: "React",
      description: "Uma das bibliotecas mais utilizadas da atualidade.",
    },
    {
      id: 3,
      title: "CESAR",
      description: "Centro de Estudos e Sistemas Avançados do Recife.",
    },
  ];

  beforeEach(() => {
    vitest.restoreAllMocks();
  });

  it("renders loading state", async () => {
    vi.spyOn(notes, "useNotes").mockReturnValue({
      notes: [],
      isPending: true,
      isError: false,
    });

    const { container } = render(
      <ProviderWrapper>
        <Notebook />
      </ProviderWrapper>,
    );

    await waitFor(() => {
      expect(screen.getByText("Carregando...")).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it("renders error state", async () => {
    vi.spyOn(notes, "useNotes").mockReturnValue({
      notes: [],
      isPending: false,
      isError: true,
    });

    const { container } = render(
      <ProviderWrapper>
        <Notebook />
      </ProviderWrapper>,
    );

    await waitFor(() => {
      expect(
        screen.getByText("Que pena, algo de errado aconteceu :("),
      ).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it("renders notes correctly", async () => {
    const { container } = render(
      <ProviderWrapper>
        <Notebook />
      </ProviderWrapper>,
    );

    //expect(container).toMatchSnapshot();

    await waitFor(() => {
      expect(screen.getByText(json.notes[0].title)).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it("handles note creation", async () => {
    const formDataResponse = {
      title: "new note title",
      description: "new description",
    } satisfies NoteFormData;
    
    const { container } = render(
      <ProviderWrapper>
        <Notebook />
      </ProviderWrapper>,
    );

    await userEvent.click(screen.getByText("+"));

    const titleInput = screen.getByPlaceholderText("Title");
    expect(titleInput).toBeInTheDocument();
  
    await userEvent.type(titleInput, "new note title");
    expect(titleInput).toHaveValue("new note title");
  
    const descriptionInput = screen.getByPlaceholderText("Description");
    expect(descriptionInput).toBeInTheDocument();
  
    await userEvent.type(descriptionInput, "new description");
    expect(descriptionInput).toHaveValue("new description");
  
    await userEvent.click(screen.getByText("Salvar"));

    await waitFor(() => {
      expect(screen.getByText(formDataResponse.title)).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it("handles note editing", async () => {
    const { container } = render(
      <ProviderWrapper>
        <Notebook />
      </ProviderWrapper>,
    );

    const firstEditButton = screen.getAllByText("Editar")[0];


    await userEvent.click(firstEditButton);

    const titleInput = screen.getByPlaceholderText("Title");
    expect(titleInput).toBeInTheDocument();
  
    await userEvent.clear(titleInput);
    await userEvent.type(titleInput, "edited note title");
  
    const descriptionInput = screen.getByPlaceholderText("Description");
    expect(descriptionInput).toBeInTheDocument();
    
    await userEvent.clear(descriptionInput);
    await userEvent.type(descriptionInput, "edited description");
  
    await userEvent.click(screen.getByText("Salvar"));

    await waitFor(() => {
      expect(screen.getByText("edited note title")).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it("handles note deletion", async () => {
    const { container } = render(
      <ProviderWrapper>
        <Notebook />
      </ProviderWrapper>,
    );

    const firstTitle = screen.getByText("Lembrete");
    const firstDeleteButton = screen.getAllByText("Remover")[0];

    await userEvent.click(firstDeleteButton);

    // Wait for mutation to complete
    await waitFor(() => {
      expect(firstTitle).not.toBeInTheDocument();
    });
  });

  /*
  

  it("handles note editing", async () => {
    render(Notebook);

    fireEvent.click(screen.getByText("Editar"));

    // Simulate form input
    fireEvent.input(screen.getByPlaceholderText("Título"), {
      target: { value: "Updated Note Title" },
    });
    fireEvent.input(screen.getByPlaceholderText("Descrição"), {
      target: { value: "Updated Note Description" },
    });

    fireEvent.click(screen.getByText("Salvar"));

    // Wait for mutation to complete
    await waitFor(() => {
      expect(mockEditMutation).toHaveBeenCalledWith({
        id: mockNotes[0].id,
        title: "Updated Note Title",
        description: "Updated Note Description",
      });
    });
  });

  it("handles note deletion", async () => {
    render(Notebook);

    fireEvent.click(screen.getByText("Remover"));

    // Wait for mutation to complete
    await waitFor(() => {
      expect(mockDeleteMutation).toHaveBeenCalledWith(mockNotes[0].id);
    });
  });*/
});
