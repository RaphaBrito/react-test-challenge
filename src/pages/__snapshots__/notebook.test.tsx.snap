// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Notebook component > handles note creation 1`] = `
<div>
  <div>
    <h1>
      Bloco de notas
    </h1>
    <div
      class="_listGrid_f34e7c"
    >
      <div
        class="_card_277467"
      >
        <h3>
          Lembrete
        </h3>
        <p>
          Colocar presença na aula.
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <div
        class="_card_277467"
      >
        <h3>
          React
        </h3>
        <p>
          Uma das bibliotecas mais utilizadas da atualidade.
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <div
        class="_card_277467"
      >
        <h3>
          CESAR
        </h3>
        <p>
          Centro de Estudos e Sistemas Avançados do Recife.
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <div
        class="_card_277467"
      >
        <h3>
          new note title
        </h3>
        <p>
          new description
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <button>
        +
      </button>
    </div>
  </div>
</div>
`;

exports[`Notebook component > handles note editing 1`] = `
<div>
  <div>
    <h1>
      Bloco de notas
    </h1>
    <div
      class="_listGrid_f34e7c"
    >
      <div
        class="_card_277467"
      >
        <h3>
          edited note title
        </h3>
        <p>
          edited description
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <div
        class="_card_277467"
      >
        <h3>
          React
        </h3>
        <p>
          Uma das bibliotecas mais utilizadas da atualidade.
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <div
        class="_card_277467"
      >
        <h3>
          CESAR
        </h3>
        <p>
          Centro de Estudos e Sistemas Avançados do Recife.
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <div
        class="_card_277467"
      >
        <h3>
          new note title
        </h3>
        <p>
          new description
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <button>
        +
      </button>
    </div>
  </div>
</div>
`;

exports[`Notebook component > renders error state 1`] = `
<div>
  <h1>
    Que pena, algo de errado aconteceu :(
  </h1>
</div>
`;

exports[`Notebook component > renders loading state 1`] = `
<div>
  <h1>
    Carregando...
  </h1>
</div>
`;

exports[`Notebook component > renders notes correctly 1`] = `
<div>
  <div>
    <h1>
      Bloco de notas
    </h1>
    <div
      class="_listGrid_f34e7c"
    >
      <div
        class="_card_277467"
      >
        <h3>
          Lembrete
        </h3>
        <p>
          Colocar presença na aula.
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <div
        class="_card_277467"
      >
        <h3>
          React
        </h3>
        <p>
          Uma das bibliotecas mais utilizadas da atualidade.
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <div
        class="_card_277467"
      >
        <h3>
          CESAR
        </h3>
        <p>
          Centro de Estudos e Sistemas Avançados do Recife.
        </p>
        <div
          class="_cardFooter_277467"
        >
          <button>
            Editar
          </button>
          <button>
            Remover
          </button>
        </div>
      </div>
      <button>
        +
      </button>
    </div>
  </div>
</div>
`;
